libcylinder/src/lib.rs

```diff
@@ -15,5 +16,30 @@
  * ------------------------------------------------------------------------------
  */
 
+mod error;
 mod hex;
+mod key;
 pub mod signing;
+
+pub use error::{SignatureVerificationError, SigningError};
+pub use key::{PrivateKey, PublicKey};
+
+/// A signer for arbitrary messages
+pub trait Signer: Send {
+    /// Signs the given message
+    fn sign(&self, message: &[u8]) -> Result<Vec<u8>, SigningError>;
+
+    /// Returns the signer's public key
+    fn public_key(&self) -> Result<PublicKey, SigningError>;
```

Comment:
This should return Result<&PublicKey, SIgningError> so we aren't forcing a clone. The first time we generate the key, we should cache it within the Signer and return the same thing if multiple public_key() calls are made.

Reply:
Updated

Reply:
This is actually quite hard to do in practice, given how safe interior mutability works.  It is probably doable, but we'd be forcing all implementations to use unsafe interior mutability.